torch/distributed/fsdp/_init_utils.py

```diff
@@ -847,37 +856,69 @@ def _need_to_materialize_module(
 
 
 def _materialize_with_param_init_fn(
-    module: nn.Module,
-    param_init_fn,
+    root_module: nn.Module,
+    param_init_fn: Callable[[nn.Module], None],
 ) -> None:
     if not callable(param_init_fn):
         raise ValueError(
             f"Expected {param_init_fn} to be callable but got {type(param_init_fn)}"
         )
-    param_init_fn(module)
+    modules_to_materialize = _get_modules_to_materialize(root_module)
+    for module in modules_to_materialize:
+        param_init_fn(module)
 
 
 def _materialize_meta_module(
-    module: nn.Module,
+    root_module: nn.Module,
     device_from_device_id: Optional[torch.device],
 ):
     # Run default meta device initialization
     materialization_device = device_from_device_id or torch.device(
         torch.cuda.current_device()
     )
-    module.to_empty(device=materialization_device)
+    modules_to_materialize = _get_modules_to_materialize(root_module)
     try:
+        # Assume that each module's `reset_parameters()` only initializes its
+        # own parameters and not those of its children
         with torch.no_grad():
-            module.reset_parameters()  # type: ignore[operator]
+            for module in modules_to_materialize:
+                # As a contract to the user, only call `reset_parameters()` if
+                # the module has directly managed parameters/buffers
+                has_module_states = (
+                    len(list(module.parameters(recurse=False))) > 0
+                    or len(list(module.buffers(recurse=False))) > 0
+                )
+                if has_module_states:
+                    module.to_empty(device=materialization_device, recurse=False)
+                    module.reset_parameters()  # type: ignore[operator]
     except BaseException as e:
         warnings.warn(
             "Unable to call `reset_parameters()` for module on meta "
-            f"device with error {str(e)}. Please ensure your "
-            "module implements a `reset_parameters()` method."
+            f"device with error {str(e)}. Please ensure that your module of"
+            f"type {type(module)} implements a `reset_parameters()` method."
```

Comment:
this BaseException was before this PR, but shouldn't we be more specific and except a NotImplementedError? Seems like this will catch all exceptions and maybe report a misleading warning if there was some other issue

Reply:
I will look into this in a follow-up.